### PR TITLE
Allow developer to add custom sound constructor

### PR DESCRIPTION
**Pull request by cstoquer**

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -78,9 +82,23 @@ AudioManager.prototype.init = function () {
...
+ * @param {string} channelId - channel id where the sound will be used
+ * @param {string} soundId   - sound id
+ *
+ * @return {Object} sound object
+ */
+AudioManager.prototype.getEmptySound = function (channelId, soundId) {
 	var sound;
+
+	// custom sound constructor
+	var constructor = this.settings.getSoundConstructor(channelId, id);
```
Comment (tbrebant) [inline index.js]: `soundId` instead of `id` I think ;)

File: index.js [hunk 2]
```diff
@@ -78,9 +82,23 @@ AudioManager.prototype.init = function () {
...
+	if (constructor) {
+		sound = new constructor();
```
Comment (tbrebant) [inline index.js]: Maybe we should add
```
if (constructor.prototype.audioManager !== this) {
	constructor.prototype.audioManager = this;
}
```
to be sure the constructor can work exactly as a regular `SoundObject`?

Comment (ronkorving) [inline index.js]: That seems very unsafe. What if AudioManager is multiple levels up the prototype chain. I would rather you not mess with my class. If you just wanna check if the prototype is there at all, there are functions and operators for that (eg: `sound instanceof AudioManager`).

Comment (tbrebant) [inline index.js]: My comment has been invalidated by the latest PR:
https://github.com/Wizcorp/AudioManager/pull/23/commits/f9f66594148b3833d121f12e595968951c43ea36
;)

**Comment by cstoquer** [comment]
updated

**Comment by tbrebant** [comment]
LGTM and tested, good to merge!

**Comment by tbrebant** [comment]
Thanks for the merge 😉

**Comment by tbrebant** [comment]
You are going to bump the version to 0.1.14 right?

**Comment by cstoquer** [comment]
@tbrebant yes the version have been bumped to 0.1.14

**Comment by tbrebant** [comment]
On npm it seems to still be 0.1.13. Is it taking time to be updated?

File: ISound.js [hunk 3]
```diff
@@ -181,3 +147,36 @@ ISound.prototype.stop = function (cb) {
 	this.playing = false;
 	return cb && cb();
 };
+
+//▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄
+/** Load sound. Abstract method to be overwritten
+ * @private
+ */
+ISound.prototype._load = function () {
+	console.log('ISound load call: ' + this.id);
```
Comment (jperdereau) [inline ISound.js]: do you need the console.log?

Comment (cstoquer) [inline ISound.js]: This is an abstract class. This method has to be override in the implementation.

Comment (ronkorving) [inline ISound.js]: Throw instead?

Comment (cstoquer) [inline ISound.js]: there is a fallback to this class if neither WebAudio nor HTML audio is available.
I don't think we should Throw in that case.

**Pull request merged**